### PR TITLE
GAWB-931 Split up transactions in submission and workflow monitor

### PR DESCRIPTION
**Pull request by helgridly**
As per the email thread re deadlocks in sharded Cromwell.
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] **Submitter**: If PR includes new or changed db queries, include the explain plans in the description
- [x] **Submitter**: Update FISMA documentation if changes to:
  - Authentication
  - Authorization
  - Encryption
  - Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell  that the PR exists if he wants to look at it **(apply requires_doge label)**
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- Review cycle:
  - LR reviews
  - Rest of team may comment on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH
  - Submitter updates documentation as needed
  - Submitter **reassigns to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 1]
```diff
@@ -244,7 +263,10 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging {
         _ <- saveWorkspace(dataAccess, updatedEntitiesAndWorkspace)
         _ <- saveEntities(dataAccess, workspace, updatedEntitiesAndWorkspace)
         _ <- saveErrors(updatedEntitiesAndWorkspace.collect { case Right(errors) => errors }, dataAccess)
-      } yield Unit
+      } yield Unit)  map (_ => Unit)
+      //map to Unit? apparently so, otherwise scala thinks the return type of this function is:
+      //DBIOAction[Unit.type, NoStream, Read with Read with Read with Read with Write with Read with Write with Read with Write with Write]
+      //...not what we mean.
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala]: just to be extra clear here: this insane generated type is not compatible with `ReadWriteAction[Unit]`, either. It's its own monster.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 2]
```diff
@@ -96,20 +96,25 @@ trait WorkflowSubmission extends FutureSupport with LazyLogging with MethodWiths
...
-      // submit the batch we found
-      case workflowRecs if workflowRecs.nonEmpty => SubmitWorkflowBatch(workflowRecs.map(_.id))
-
-      // if we didn't find any then look again in a little bit
-      case _ => ScheduleNextWorkflowQuery
+    //flip the workflows to Launching in a separate txn.
+    //if this optimistic-lock-exceptions with another txn, this one will barf and we'll reschedule when we pipe it back to ourselves
+    workflowRecsToLaunch map { wfRecs =>
+      dataSource.inTransaction { dataAccess =>
+        dataAccess.workflowQuery.batchUpdateStatus(wfRecs, WorkflowStatuses.Launching).map(_ => wfRecs)
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: this has been moved up here from inside `reserveWorkflowBatch`, if you were looking for it

**Comment by helgridly** [comment]
hmm. according to [the logs in the failed test](https://travis-ci.org/broadinstitute/rawls/builds/150169193#L2569):

`[WARN] [22:10:45.168] [test-thread-132976] o.b.d.r.j.WorkflowSubmissionSpec$TestWorkflowSubmission - There are 9 active workflows which is beyond the total active cap of 0. Workflows will not be submitted.`

Except the workflow submission actor [made by the test](https://github.com/broadinstitute/rawls/blob/split_txns/src/test/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionSpec.scala#L74) has a cap of 100. I am not sure if this is a clue or if it's just out-of-order logging (given that the next test tests a cap of 0).

There's also [this](https://travis-ci.org/broadinstitute/rawls/builds/150169193#L2653):

`[ERROR] [22:10:51.037] [WorkflowSubmissionSpec-akka.actor.default-dispatcher-7] o.b.d.r.j.WorkflowSubmissionActor - could not update 1 workflows because their record version(s) have changed`

...which may indicate that we're doing some nondeterministic stomping.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala [hunk 3]
```diff
@@ -35,7 +37,8 @@ object SubmissionMonitorActor {
 
   /**
    * The response from querying the exec services.
-   * @param statusResponse If a successful response shows an unchanged status there
+    *
+    * @param statusResponse If a successful response shows an unchanged status there
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:41]: accidental spacing changes here and with the two other params comments below?

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala:41]: my intellij insists on alphabetising imports and making doc comments SUPER LEGIT and i don't know how to turn it off. for the record, i am not a horrendous pedant (in this particular case)

**Comment by MatthewBemis** [comment]
:shipit:

**Pull request merged**